c7n/resolver.py

```diff
@@ -157,18 +166,54 @@ def get_contents(self):
         return contents, format
 
     def get_values(self):
-        key = [self.data.get(i) for i in ('url', 'format', 'expr', 'headers')]
+        cache_key = [self.data.get(i) for i in ('url', 'format', 'expr', 'headers', 'query')]
         with self.cache:
             # use these values as a key to cache the result so if we have
             # the same filter happening across many resources, we can reuse
             # the results.
-            contents = self.cache.get(("value-from", key))
+            contents = self.cache.get(("value-from", cache_key))
             if contents is not None:
                 return contents
-            contents = self._get_values()
-            self.cache.save(("value-from", key), contents)
+            if self.data['url'] == 'dynamodb':
+                contents = self._get_ddb_values()
+            else:
+                contents = self._get_values()
+            self.cache.save(("value-from", cache_key), contents)
             return contents
 
+    def _get_ddb_values(self):
+        if not self.data['query']:
+            return
+        if not self.data['query'].lower().startswith('select'):
+            return
+
+        from boto3.dynamodb.types import TypeDeserializer
+        from botocore.paginate import Paginator
+
+        client = local_session(self.manager.session_factory).client('dynamodb')
+
+        pager = Paginator(
+            client.execute_statement,
+            {"input_token": "NextToken", "output_token": "NextToken", "result_key": "Items"},
+            client.meta.service_model.operation_model('ExecuteStatement')
+        )
+        deserializer = TypeDeserializer()
+        results = []
+
+        record_singleton = False
+        # result = client.execute_statement(Statement=self.data['query'])
+        for page in pager.paginate(Statement=self.data['query']):
+            for row in page.get("Items", []):
+                record = {k: deserializer.deserialize(v) for k, v in row.items()}
+                if record_singleton or len(record) == 1:
+                    record_singleton = True
+                    results.append(list(record.values())[0])
+                else:
+                    results.append(record)
+        if not record_singleton and self.data.get('expr'):
```

Comment:
I was testing this locally with a couple different ddb setups. I noticed that if I had a table item like this:

```json
{
  "pk": {
    "S": "ec2-image-details"
  },
  "resource_ids": {
    "L": [
      {
        "S": "i-00d3f85ef9622b0e2"
      }
    ]
  }
}
```

and a filter like this:

```yaml
      - type: value
        key: InstanceId
        op: not-in
        value_from:
          url: dynamodb
          query: |
            select resource_ids from c7n_exceptions
            where pk = '{policy[name]}'
          expr: '[]'
          format: json
```

Then the value_from lookup would return a list of lists. The `expr` wasn't getting evaluated because it came across as `record_singleton = True`. Making this change got things working:

```suggestion
        if not record_singleton or self.data.get('expr'):
```

It _feels_ like a reasonable use case to have table items contain a list of resource IDs that then need flattening, but there may be a better way to handle that case. Worth noting that having a table item for each resource works without the change, and I'm sure there are other ways to make multiple resources per item work too...

Reply:
that seems okay, although I'm a little less clear on the best intent for doing multiple values in a single item, but I guess 400k row size does allow some flexibility.